src/DssGateSuck.sol

```diff
@@ -0,0 +1,171 @@
+        _;
+    }
+
+    // --- Integration Access Control ---
+    mapping (address => uint256) public bud;
+    event Kiss(address a);
+    event Diss(address a);
+    function kiss(address _a) external auth {
+        require(_a != address(0), "bud/no-contract-0");
+        require(bud[_a] == 0, "bud/approved");
+        bud[_a] = 1;
+        emit Kiss(_a);
+    }
+    function diss(address _a) external auth {
+        require(bud[_a] == 1, "bud/not-approved");
+        bud[_a] = 0;
+        emit Diss(_a);
+    }
+    modifier toll { require(bud[msg.sender] == 1, "bud/not-authorized"); _; }
+
+    // --- Math ---
+    function _add(uint x, uint y) internal pure returns (uint z) {
+        require((z = x + y) >= x);
+    }
+    function _sub(uint x, uint y) internal pure returns (uint z) {
+        require((z = x - y) <= x);
+    }
+
+    /// maker protocol vat
+    address public immutable vat;
+    /// maker protocol vow
+    address public immutable vow;
+
+    /// draw limit- total amount that can be drawn from vat.suck
+    uint256 public max; // [rad]
+    // amount drawn- amount currently drawn and not put back
+    uint256 public fill = 0;
+
+    constructor(address vat_, address vow_) public {
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+
+        vat = vat_; // set vat address
+        vow = vow_; // set vow address
+    }
+
+    // --- Events ---
+    event File(bytes32 indexed what, uint256 data);
+    event Draw(address indexed dst, uint256 amount); // log upon draw
+
+    // --- UTILS ---
+    function _max(uint x, uint y) internal pure returns (uint z) {
+        return x <= y ? y : x;
+    }
+
+    // --- Draw Limits ---
+    /// Update draw limit
+    /// @dev Restricted to authorized governance addresses
+    /// @dev Approved total can be updated to both a higher or lower value
+    /// @param what what value are we updating
+    /// @param data what are we updating it to
+    function file(bytes32 what, uint256 data) external auth {
+        if (what == "max") {
+            max = data; // update approved total amount
+            emit File(what, data);
+        }
+    }
+
+    /// Suck to destination
+    /// @dev Restricted to approved integration addresses
+    /// @param dst who are you sucking to
+    /// @param amt dai amount in rad
+    function suck(address dst, uint256 amt) public toll returns (bool) {
+        require(VatLike(vat).live() == 1, "dss-gate/vat-not-live");
+
+        fill = _add(fill, amt);
+        require(max >= fill, "dss-gate/insufficient-allowance");
+
+        VatLike(vat).suck(vow, dst, amt);
+        emit Draw(dst, amt);
+    }
+
+    /// Suck to sender
+    /// @dev Restricted to approved integration addresses (because it calls the public suck)
+    /// @param amt dai amount in rad
+    function suck(uint256 amt) external {
+        suck(msg.sender, amt);
+    }
+
+    /// Repay dai (reverse suck)
+    function blow(uint256 amt) external {
+        if (fill >= amt) {
```

Comment:
Not sure I understand this.
If someone returns more amount than the outstanding dai (`fill`) then we don't regard it as returned dai in the `fill` accounting but do not revert?

Reply:
basically, if they are repaying we want to decrease fill but would still want to allow them to return Dai to the system if it is above the amount they took out (i.e. they generated profit).does that make sense?

Reply:
mm what still isn't clear to me is why if they pay back `amt` which is equal to `fill` (or smaller) we do account for that but they also pay back some additional profit then we disregard that completely and don't decrease `fill`

Reply:
how do you feel about this: https://github.com/makerdao/dss-gate/pull/1/commits/a1a8d1458168f24442a1580fae3fffa83159facc ?